File: Mac/TabletPenMac/PenData.swift
import Foundation

/// Represents pen input data received from the Android tablet.
struct PenData {
    let x: Float           // Normalized X position (0.0 to 1.0)
    let y: Float           // Normalized Y position (0.0 to 1.0)
    let pressure: Float    // Pressure (0.0 to 1.0)
    let isHovering: Bool   // True when pen is near but not touching
    let isDown: Bool       // True when pen is touching screen
    let buttonPressed: Bool // True when pen button is pressed
    let tiltX: Float       // Tilt on X axis
    let tiltY: Float       // Tilt on Y axis
    let timestamp: Int64   // Event timestamp

    /// Parse pen data from the network format.
    /// Format: "x,y,pressure,hovering,down,button,tiltX,tiltY,timestamp"
    static func parse(from string: String) -> PenData? {
        let parts = string.trimmingCharacters(in: .whitespacesAndNewlines).split(separator: ",")
        guard parts.count >= 9 else { return nil }

        guard let x = Float(parts[0]),
              let y = Float(parts[1]),
              let pressure = Float(parts[2]),
              let tiltX = Float(parts[6]),
              let tiltY = Float(parts[7]),
              let timestamp = Int64(parts[8]) else {
            return nil
        }

        return PenData(
            x: x,
            y: y,
            pressure: pressure,
            isHovering: parts[3] == "1",
            isDown: parts[4] == "1",
            buttonPressed: parts[5] == "1",
            tiltX: tiltX,
            tiltY: tiltY,
            timestamp: timestamp
        )
    }
}

File: Android/app/src/main/java/com/example/tabletpen/PenData.kt
package com.example.tabletpen

/**
 * Data class representing pen input state.
 * This gets serialized and sent to the Mac.
 */
data class PenData(
    val x: Float,           // Normalized X position (0.0 to 1.0)
    val y: Float,           // Normalized Y position (0.0 to 1.0)
    val pressure: Float,    // Pressure (0.0 to 1.0), 0 when hovering
    val isHovering: Boolean,// True when pen is near but not touching
    val isDown: Boolean,    // True when pen is touching screen
    val buttonPressed: Boolean, // True when pen button is pressed
    val tiltX: Float,       // Tilt on X axis (-1.0 to 1.0)
    val tiltY: Float,       // Tilt on Y axis (-1.0 to 1.0)
    val timestamp: Long     // Event timestamp
) {
    /**
     * Serialize to a simple string format for network transmission.
     * Format: "x,y,pressure,hovering,down,button,tiltX,tiltY,timestamp"
     */
    fun serialize(): String {
        return buildString {
            append("%.5f".format(x))
            append(",")
            append("%.5f".format(y))
            append(",")
            append("%.4f".format(pressure))
            append(",")
            append(if (isHovering) "1" else "0")
            append(",")
            append(if (isDown) "1" else "0")
            append(",")
            append(if (buttonPressed) "1" else "0")
            append(",")
            append("%.4f".format(tiltX))
            append(",")
            append("%.4f".format(tiltY))
            append(",")
            append(timestamp)
            append("\n")
        }
    }

    companion object {
        /**
         * Deserialize from string format.
         */
        fun deserialize(data: String): PenData? {
            return try {
                val parts = data.trim().split(",")
                if (parts.size >= 9) {
                    PenData(
                        x = parts[0].toFloat(),
                        y = parts[1].toFloat(),
                        pressure = parts[2].toFloat(),
                        isHovering = parts[3] == "1",
                        isDown = parts[4] == "1",
                        buttonPressed = parts[5] == "1",
                        tiltX = parts[6].toFloat(),
                        tiltY = parts[7].toFloat(),
                        timestamp = parts[8].toLong()
                    )
                } else null
            } catch (e: Exception) {
                null
            }
        }
    }
}
